src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -206,6 +206,8 @@ private static async ValueTask<SslStream> EstablishSslConnectionAsyncCore(bool a
                     throw CancellationHelper.CreateOperationCanceledException(e, cancellationToken);
                 }
 
+                // ToDo: let this exception buuble up to the caller, or should we somehow recognize that server didn't allow
+                //       particular HTTP version in ALPN (how? Furtik will know) and use NSE instead.
```

Comment:
Yea I think we need an exception message saying if ALPN was the reason for failure. @wfurt will know how to check this.

Reply:
SslStream.NegotiatedApplicationProtocol should be SslApplicationProtocol.Http2 when TLS handshake is completed. There is example in GetHttp2ConnectionAsync()

Reply:
@wfurt: this is an exception handling for `AuthenticateAsClient`, will `SslStream.NegotiatedApplicationProtocol` have anything meaningful if negotiation failed?
Can `AuthenticateAsClient` fail for anything else than failed ALPN due to server not supporting advertised protocol? I'd guess yes. If so, how can I distinguish such error (server not supporting protocol I want) from all others?

Reply:
oh, I did not quite review the whole PR. And yes, SslStream can fail for unrelated reasons and there is no crips way how to know what exactly happened. On mismatch, server can simply close connection and we would see IO exception or it can sent alert. There is not one dedicated to ALPN so we would just see generic AuthenticationException.
I think  we should just bubble up exception, possibly wrapper in HTTP one.

Reply:
That's what we already do, so I leave it as it is now. We can improve the exception anytime in the future if we get any feedback.

Reply:
I was thinking about it more, and I think that is the right options. I did not see any tracing added. Perhaps that is something you should do so we could get record of requests policy.

Reply:
[RFC](https://tools.ietf.org/html/rfc7301#section-3.2) says:
> In the event that the server supports no
   protocols that the client advertises, then the server SHALL respond
   with a fatal "no_application_protocol" alert.

I'll look if we somehow have a status/error code for this. If so, at least this one could be handled separately.

Reply:
Bah, it seems it's not followed by servers:
https://bugs.chromium.org/p/chromium/issues/detail?id=497770
https://mta.openssl.org/pipermail/openssl-dev/2015-June/001838.html

Keeping the connection and letting client fallback to HTTP/1.1